### DISCUSSION TITLE
many times, we only need to change the width of the current beat

### DISCUSSION THREAD

**Discussion by goodgame365** [CoderLine/alphaTab#1654] [General]
think the method this.addGlyph(new SpacingGlyph(0, 0, 4 * this.scale)); changes the width of all the beats inside the entire bar. However, many times, we only need to change the width of the current beat, for example, when there are grace notes or accidentals in front of it, while other beats do not require widening.

![未标题-1](https://github.com/user-attachments/assets/1e4087bd-bdf9-40c1-a299-42eb4ce82373)

**Comment by Danielku15** [comment]
**Generally:** 

alphaTab is using an adapted version of "Gourlays' Algorithm For Spacing" for placing music notation. See [4.5.1 Spacing](https://guido.grame.fr/papers/kai_renz_diss.pdf). 

Every note defines what exact space it needs as minimum to place its glyphs and then the notes are aligned in a way that they are correctly placed to the time when they are played. If a note has accidentals (or any glyph like spacings) this is registered as "minimum space needed". Hence all glyphs influence in some way the spacing and alignment. 

But assuming there is enough space alphaTab should align notes as expected not scaling unnecessarily bigger (see below).

**On your screenshot**
I don't see a problem in your screenshot. You mention that bar 6 is correct. According to the defined spacing and sizes it requires more than 50% of the available width and therefore it can only place 1 bar per line. Then accordingly to the algorithm the bar is stretched to the full width keeping the alignments correct using the "spring" influence. 

The general size is calculated by the Gourlay algorithm respecting the "stretch force" option of alphaTab. The stretch force basically describes how strong the "springs" are and how wide the bars are stretching due to this.

I agree that maybe sometimes in alphaTab the exact spaces/paddings between the glyph are not very optimized. The general glyph sizing, alignments and paddings are somthing which could use a larger refactoring but generally they should be correct. Beats will receive widening as the bar is stretched to ensure correct placement. 

Here a example showing how spacing works in some different products

AlphaTab: we nicely scale the bars without unnecessary spacing keeping the timely alignment if possible. 
![image](https://github.com/user-attachments/assets/532d77f4-625d-4e27-a478-e64098cfe94b)

Guitar Pro 8: it scales the bar with accidentals slightly bigger. 
![image](https://github.com/user-attachments/assets/61792f46-9a09-4fd2-ae55-22792328bf6c)

MuseScore 4: like alphaTab it keeps the alignment on beat times if possible (no Jianpu but you get the idea):
![image](https://github.com/user-attachments/assets/ab4709c5-eab4-414a-96a9-93f45cea8d26)

I hope I was able to cover your question with this.

**Comment by goodgame365** [comment]
However, I still feel it's a bit strange. The effect may not be obvious with a single accidental, but if multiple grace notes are connected together, it makes the width of the other beats without grace notes appear particularly wide, as shown in the illustration.

This is still secondary; the main issue is that we cannot predict the width of the bar, nor can we accurately determine how much the width value will expand with each additional grace note.
![1](https://github.com/user-attachments/assets/c896f7aa-b56a-45c2-8d55-24e2c5d9aaaa)
![未标题-2](https://github.com/user-attachments/assets/e88057f6-e413-4ddd-8bd1-2a0ed46e57fb)
Guitar Pro 8:
![未标题-3](https://github.com/user-attachments/assets/a95dd97a-ec4f-409a-a7f8-a4b991968223)
